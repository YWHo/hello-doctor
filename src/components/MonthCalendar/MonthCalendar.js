import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import dayjs from 'dayjs';
import PropTypes from 'prop-types';
import { getSelectedDate, getShowingCalendar } from '../../state/selectors';
import { saveSelectedDate, toggleShowingCalendar } from '../../state/actions';
import { getWeekdayLabels, getMonthArray2dOf } from '../../shared/dateHelper';

const Container = styled.div`
  display: block;
  background-color: #177d91;
  padding-top: ${(props) => (props.show ? '20px' : 0)};
  padding-bottom: ${(props) => (props.show ? '16px' : 0)};
  height: 100%;
  max-height: ${(props) => (props.show ? '328px' : 0)};
  transition:
    max-height 1s,
    padding-top 1s,
    padding-bottom 1s;
  transition-delay: ${(props) => (props.show ? 0 : '1s')};
`;

const TableView = styled.table`
  margin: auto;
  visibility: ${(props) => (props.show ? 'visible' : 'hidden')};
  opacity: ${(props) => (props.show ? 1 : 0)};
  transition:
    visibility 1.5s,
    opacity 1.5s;
  transition-delay: ${(props) => (props.show ? '0.5s' : 0)};
`;

const TableData = styled.td`
  height: 40px;
  width: 48px;
`;

const DateLabel = styled.div`
  background-color: ${(props) =>
    props.selected ? '#0E6171' : props.isAdjacent ? '#147487' : 'transparent'};
  border-radius: ${(props) =>
    props.selected || props.isAdjacent ? '50%' : '100%'};
  font-family: 'Roboto', sans-serif;
  font-weight: ${(props) => (props.selected ? 500 : 300)};
  font-size: 20px;
  color: ${(props) =>
    props.isHeader || props.selected
      ? '#fff'
      : props.enabled
      ? '#bbd9dd'
      : '#8bbcc6'};
  line-height: 38px;
  height: 38px;
  width: 38px;
  text-align: center;
  vertical-align: middle;
`;

export function MonthCalendar(props) {
  const { isShowingCalendar = false } = props;

  return (
    <Container show={isShowingCalendar}>
      <TableView show={isShowingCalendar}>
        <tbody>
          {showWeekLabels()}
          {showCalendarRows(props)}
        </tbody>
      </TableView>
    </Container>
  );
}

function showWeekLabels() {
  const weekLabels = getWeekdayLabels().map((name, idx) => {
    return (
      <TableData key={`td0_${idx}`}>
        <DateLabel isHeader>{name}</DateLabel>
      </TableData>
    );
  });
  return <tr>{weekLabels}</tr>;
}

function showCalendarRows(props) {
  const { selectedDate, today = dayjs() } = props;
  const adjacent = getAdjacentDateDict(selectedDate);

  return getMonthArray2dOf(selectedDate).map((row, idxA) => {
    const singleRow = row.map((dateStr, idxB) => {
      const selected = dayjs(dateStr).isSame(dayjs(selectedDate), 'day');
      const enabled =
        dateStr.length > 0 && !dayjs(dateStr).isBefore(dayjs(today), 'day');
      const isAdjacent = adjacent[dayjs(dateStr).format('YYYY-MM-DD')];
      const dateVal = dateStr.length > 0 ? dayjs(dateStr).format('D') : '';
      return (
        <TableData key={`tc_${idxA}_${idxB}`}>
          <DateLabel
            selected={selected}
            isAdjacent={isAdjacent}
            enabled={enabled}
            onClick={() => onDateSelected(props, dateStr, enabled)}
          >
            {dateVal}
          </DateLabel>
        </TableData>
      );
    });

    return <tr key={`tr_${idxA}`}>{singleRow}</tr>;
  });
}

function onDateSelected(props, dateStr, enabled) {
  const { dSaveSelectedDate, dToggleShowingCalendar, isShowingCalendar } =
    props;
  if (!enabled) return;

  const newDateStr = dayjs(dateStr).format('YYYY-MM-DD');
  dSaveSelectedDate(dayjs(newDateStr));

  // Using setTimeout as a workaround to solve the problem of immediately
  // hide calendar due to CSS animation failed to work for an unknown reason
  setTimeout(() => dToggleShowingCalendar(!isShowingCalendar), 100);
}

function getAdjacentDateDict(refDate) {
  const adjacent = {};
  [-2, -1, 1, 2].forEach((val) => {
    const dateStr = dayjs(refDate).add(val, 'day').format('YYYY-MM-DD');
    adjacent[dateStr] = true;
  });
  return adjacent;
}

MonthCalendar.propTypes = {
  selectedDate: PropTypes.object.isRequired,
  today: PropTypes.object,
};

const mapStateToProps = (state) => {
  return {
    isShowingCalendar: getShowingCalendar(state),
    selectedDate: getSelectedDate(state),
  };
};

function mapDispatchToProps(dispatch) {
  return {
    dSaveSelectedDate: (value) => dispatch(saveSelectedDate(value)),
    dToggleShowingCalendar: (status) => dispatch(toggleShowingCalendar(status)),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(MonthCalendar);
