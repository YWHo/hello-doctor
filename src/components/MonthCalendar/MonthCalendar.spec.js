import React from 'react';
import dayjs from 'dayjs';
import { shallow } from 'enzyme';
import { MonthCalendar } from './MonthCalendar';

describe('MonthCalendar', () => {
  it('render correctly', () => {
    const wrapper = shallow(
      <MonthCalendar
        selectedDate={dayjs('2019-06-27')}
        today={dayjs('2019-06-23')}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
