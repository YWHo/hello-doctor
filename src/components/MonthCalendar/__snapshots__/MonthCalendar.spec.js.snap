// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthCalendar render correctly 1`] = `
<styled.div
  $show={false}
>
  <styled.table
    $show={false}
  >
    <tbody>
      <tr>
        <styled.td
          key="td0_0"
        >
          <styled.div
            $isHeader={true}
          >
            S
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_1"
        >
          <styled.div
            $isHeader={true}
          >
            M
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_2"
        >
          <styled.div
            $isHeader={true}
          >
            T
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_3"
        >
          <styled.div
            $isHeader={true}
          >
            W
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_4"
        >
          <styled.div
            $isHeader={true}
          >
            T
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_5"
        >
          <styled.div
            $isHeader={true}
          >
            F
          </styled.div>
        </styled.td>
        <styled.td
          key="td0_6"
        >
          <styled.div
            $isHeader={true}
          >
            S
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_0"
      >
        <styled.td
          key="tc_0_0"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_1"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_2"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_3"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_4"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_5"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          />
        </styled.td>
        <styled.td
          key="tc_0_6"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            1
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_1"
      >
        <styled.td
          key="tc_1_0"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            2
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_1"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            3
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_2"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            4
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_3"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            5
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_4"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            6
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_5"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            7
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_1_6"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            8
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_2"
      >
        <styled.td
          key="tc_2_0"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            9
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_1"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            10
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_2"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            11
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_3"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            12
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_4"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            13
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_5"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            14
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_2_6"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            15
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_3"
      >
        <styled.td
          key="tc_3_0"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            16
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_1"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            17
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_2"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            18
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_3"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            19
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_4"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            20
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_5"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            21
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_3_6"
        >
          <styled.div
            $enabled={false}
            onClick={[Function]}
            selected={false}
          >
            22
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_4"
      >
        <styled.td
          key="tc_4_0"
        >
          <styled.div
            $enabled={true}
            onClick={[Function]}
            selected={false}
          >
            23
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_1"
        >
          <styled.div
            $enabled={true}
            onClick={[Function]}
            selected={false}
          >
            24
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_2"
        >
          <styled.div
            $enabled={true}
            $isAdjacent={true}
            onClick={[Function]}
            selected={false}
          >
            25
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_3"
        >
          <styled.div
            $enabled={true}
            $isAdjacent={true}
            onClick={[Function]}
            selected={false}
          >
            26
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_4"
        >
          <styled.div
            $enabled={true}
            onClick={[Function]}
            selected={true}
          >
            27
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_5"
        >
          <styled.div
            $enabled={true}
            $isAdjacent={true}
            onClick={[Function]}
            selected={false}
          >
            28
          </styled.div>
        </styled.td>
        <styled.td
          key="tc_4_6"
        >
          <styled.div
            $enabled={true}
            $isAdjacent={true}
            onClick={[Function]}
            selected={false}
          >
            29
          </styled.div>
        </styled.td>
      </tr>
      <tr
        key="tr_5"
      >
        <styled.td
          key="tc_5_0"
        >
          <styled.div
            $enabled={true}
            onClick={[Function]}
            selected={false}
          >
            30
          </styled.div>
        </styled.td>
      </tr>
    </tbody>
  </styled.table>
</styled.div>
`;
